### PR TITLE
QUIC: Default add RVCM to Envoy QUIC client sessions.

### LINKED ISSUE #16311: QUIC Connection Migration
We support QUIC upstream, but there's currently not support for migration.  I think this would have to be done in upstream Envoy even if it's likely only going to be used for mobile.

### PR DESCRIPTION
**Pull request by RenjieTang**
Signed-off-by: Renjie Tang <renjietang@google.com>

Commit Message: Default add connection option RVCM to client QUIC connections.
Additional Description: This will truly enable active port migration implemented at https://github.com/envoyproxy/envoy/pull/18103
Risk Level: Low
Testing: Unit tests
Docs Changes: n/a.
Release Notes: n/a.
Platform Specific Features: n/a.
Fixes #16311

### CODE REVIEW

File: source/common/quic/client_connection_factory_impl.cc [hunk 1]
```diff
@@ -48,6 +48,14 @@ std::unique_ptr<Network::ClientConnection> createQuicNetworkConnection(
     }
   }
 
+  // Default enable RVCM connection option so that port migration is enabled.
+  quic::QuicTagVector connection_options;
+  if (info_impl->quic_config_.HasSendConnectionOptions()) {
+    connection_options = info_impl->quic_config_.SendConnectionOptions();
+  }
+  connection_options.push_back(quic::kRVCM);
+  info_impl->quic_config_.SetConnectionOptionsToSend(connection_options);
```
Comment (danzh2010) [inline source/common/quic/client_connection_factory_impl.cc:57]: I slightly prefer to either modifying the options in PersistentQuicInfoImpl constructor or duplicating the config and modifying the copy here. Mostly because it seems unnecessary for each connection creation to make the same modification to the shared config.

Comment (danzh2010) [reply source/common/quic/client_connection_factory_impl.cc]: Nice! Can you also change quic_http_integration_test not to set --quic_remove_connection_migration_connection_option?

Comment (RenjieTang) [reply source/common/quic/client_connection_factory_impl.cc]: done.
The test quic_connection_persistent_info was not created in a kosher way.
So some more hacks were done.

File: test/mocks/upstream/cluster_info.cc [hunk 2]
```diff
@@ -61,7 +61,7 @@ MockClusterInfo::MockClusterInfo()
           runtime_, "fake_key", 1, 1024, 1024, 1, std::numeric_limits<uint64_t>::max(),
           circuit_breakers_stats_, absl::nullopt, absl::nullopt)),
       stats_scope_(stats_store_.createScope("test_scope")) {
-  ON_CALL(*this, connectTimeout()).WillByDefault(Return(std::chrono::milliseconds(1)));
+  ON_CALL(*this, connectTimeout()).WillByDefault(Return(std::chrono::milliseconds(5001)));
```
Comment (danzh2010) [inline test/mocks/upstream/cluster_info.cc:64]: default 1ms connect timeout is definitely unreasonable, but does it somehow cause the test fail?  Anyway we probably need to do sanity check on the configured value, could you file an issue about the problem you hit?

**Review by danzh2010 — DISMISSED** [review]
LGTM

**Comment by danzh2010** [comment]
Need a main sync to for the tsan failure?

File: test/mocks/upstream/cluster_info.cc (hunk 2 above)
Comment (RenjieTang) [reply test/mocks/upstream/cluster_info.cc:64]: I think this value will be converted to QUIC handshake timeout.
And I ran into the QUIC bug at https://github.com/google/quiche/blob/main/quic/core/quic_connection.cc#L4634

File: test/integration/http_integration.cc [hunk 3]
```diff
@@ -360,8 +360,9 @@ void HttpIntegrationTest::initialize() {
   registerTestServerPorts({"http"}, test_server_);
 
   // Needs to outlive all QUIC connections.
+  std::shared_ptr<Upstream::MockClusterInfo> cluster{new NiceMock<Upstream::MockClusterInfo>()};
```
Comment (RyanTheOptimist) [inline test/integration/http_integration.cc:363]: nit: Can we use make_shared here?

```
auto cluster = std::make_shared<NiceMock<Upstream::MockClusterInfo>>();
```

File: test/integration/http_integration.cc [hunk 4]
```diff
@@ -360,8 +360,9 @@ void HttpIntegrationTest::initialize() {
...
   auto quic_connection_persistent_info =
-      std::make_unique<Quic::PersistentQuicInfoImpl>(*dispatcher_, 0);
+      Quic::createPersistentQuicInfoForCluster(*dispatcher_, *cluster);
```
Comment (RyanTheOptimist) [inline test/integration/http_integration.cc:365]: Out of curiosity, why was this change needed? I'm guessing your PR exposed that the old code was doing something wrong?

**Review by RyanTheOptimist — DISMISSED** [review]
LGTM, modulo one nit and one question.

File: test/integration/http_integration.cc (hunk 4 above)
Comment (RenjieTang) [reply test/integration/http_integration.cc:365]: Yeah so normally a PersistentQuicInfo is created via this createPersistentQuicInfoForCluster() call https://github.com/envoyproxy/envoy/blob/dd141e13da7acb5b9788a46d366e3bcf6b38bd3a/source/common/http/http3/conn_pool.cc#L82, where a couple of configs are added on top of the vanilla PersistentQuicInfo.
My change of adding RVCM is one of them.

**Review by RyanTheOptimist — APPROVED** [review]

**Review by alyssawilk — APPROVED** [review]
